### DISCUSSION TITLE
RN Expo Facebook Auth Nonces Mismatch

### DISCUSSION THREAD

**Discussion by adamsucharda99** [supabase/supabase#22297] [Questions]
I'm trying to implement Facebook authentication in a React Native Expo app using Supabase and `react-native-fbsdk-next`. I'm currently following the [Google auth implementation guide](https://supabase.com/docs/guides/auth/social-login/auth-google?platform=react-native) provided by Supabase and adjusting it for Facebook, as the Supabase documentation doesn't provide clear instructions on how to implement Facebook auth with Expo. According to the Supabase API docs, the [`signInWithIdToken`](https://supabase.com/docs/reference/javascript/auth-signinwithidtoken) method accepts 'facebook' as a provider option, just like 'google'.

Here's the function I'm using for Facebook sign-in:
```
const facebookSignIn = async () => {
  try {
    const res = await LoginManager.logInWithPermissions(
      ['public_profile'],
      'limited'
    );

    if (res.isCancelled) {
      throw new Error('User cancelled the login process');
    } else if (Platform.OS === 'ios') {
      const res = await AuthenticationToken.getAuthenticationTokenIOS();

      if (res) {
        const { data, error } = await supabase.auth.signInWithIdToken({
          provider: 'facebook',
          token: res.authenticationToken,
          nonce: res.nonce,
        });

        console.log({ data, error });
      } else {
        throw new Error('No auth token present');
      }
    }
  } catch (error: any) {}
};
```

When I run this function, I get an `AuthApiError: Nonces mismatch` error. However, when I log the nonce to the console, the nonce provided to `signInWithIdToken` and the nonce I got from `getAuthenticationTokenIOS` are the same value.

I'm not sure why I'm getting this error or how to fix it. Can anyone guide me on how this login function should be done? Any help would be greatly appreciated.

**Comment by eybel** [comment]
Hello! I just started a brand new 2024 expo app and created my Development (profile) builds. I was able to see the google button and the popup google window that shows my account to login, and after I press my account it seems that it actually goes through and I see in my terminal that I get all the Google Auth info (token, email, name, etc...) but I get these errors* from the supabase.auth.signInWithIdToken({}):

*error messages:
data: {"session": null, "user": null}
error: [AuthApiError: Passed nonce and nonce in id_token should either both exist or not.]

`
import {
  GoogleSignin,
  GoogleSigninButton,
  statusCodes,
} from "@react-native-google-signin/google-signin";
import { supabase } from "@/lib/supabaseClient";

const GoogleSignInButtonProvider = () => {
  GoogleSignin.configure({
    scopes: ["https://www.googleapis.com/auth/drive.readonly"],
    webClientId:
      "XXXXXXXXXXXXXXXXXXXXXXXXX.apps.googleusercontent.com",
    iosClientId:
      "XXXXXXXXXXXXXXXXXXXXXXXXX.apps.googleusercontent.com",
  });

  const onGoogleSignInHandler = async () => {

    try {
      await GoogleSignin.hasPlayServices();

      const userInfo = await GoogleSignin.signIn();

      if (userInfo.idToken) {

        const { data, error } = await supabase.auth.signInWithIdToken({
          provider: "google",
          token: userInfo.idToken,
        });

      } else {
        throw new Error("no ID token present!");
      }
    } catch (error: any) {
      if (error.code === statusCodes.SIGN_IN_CANCELLED) {
        // user cancelled the login flow
      } else if (error.code === statusCodes.IN_PROGRESS) {
        // operation (e.g. sign in) is in progress already
      } else if (error.code === statusCodes.PLAY_SERVICES_NOT_AVAILABLE) {
        // play services not available or outdated
      } else {
        // some other error happened
      }
    }
  };
  return (
    <GoogleSigninButton
      size={GoogleSigninButton.Size.Wide}
      color={GoogleSigninButton.Color.Dark}
      onPress={onGoogleSignInHandler}
    />
  );
};

export default GoogleSignInButtonProvider;
`

A) The only way I was able to fix this is by switch the option SKIP NONCE to TRUE.... but I am not sure if I am supposed to do that, it says its less secure.
B) If I submit a random string. I get nonce mismatch
 
 
I am testing it in an IOS simulator.

**Reply by ayazalphasquad** [reply]
Hi @eybel - I'm also getting the same issue. Did you find some solution ?

**Comment by hugoh59** [comment]
Did you figure it out, I'm also struggling with this.

**Reply by eybel** [reply]
> Did you figure it out, I'm also struggling with this.

No I didnt, I just switched the SKIP NONCE to true for now. @evelant  just offered a solution thought, we could try

**Reply by easonHeyuan** [reply]
How to switch the SKIP NONCE to true, I did not found some param

**Comment by evelant** [comment]
@hugoh59 @ayazalphasquad @adamsucharda99 @eybel I figured out the problem. The issue is that gotrue (supabase auth) decodes the id token then compares the nonce from the token to _the hash of the nonce you pass in to signInWithIdToken_. That means you need to hash your nonce, pass the hashed nonce to the native sign in method, then pass the raw nonce to `signInWithIdToken`. That way gotrue will hash the raw nonce and compare it to the nonce in the token and they will match. Example:

```ts
  // Generate nonce for iOS limited login
            const nonce = Platform.OS === "ios" ? uuid() : undefined
            //need to hash the nonce for iOS limited login since supabase checks the nonce against the hashed nonce not the original
            const hashedNonce = await digestStringAsync(
                CryptoDigestAlgorithm.SHA256,
                nonce ?? "" // Provide empty string as fallback
            )
            // Request permissions with limited tracking on iOS
            const result = await LoginManager.logInWithPermissions(
                ["email", "openid"],
                Platform.OS === "ios" ? "limited" : undefined,
                // give the hashed nonce so it ends up in the token and supabase can match it by hashing the raw nonce in gotrue
                Platform.OS === "ios" ? hashedNonce : undefined
            )
            console.log(`Facebook login result`, result)
            if (result.isCancelled) {
                throw new Error("Facebook login cancelled")
            }

            // Get the authentication token
            let idToken: string
            if (Platform.OS === "ios") {
                // On iOS, use AuthenticationToken for limited login
                // This token cannot be used to access the Graph API
                const authToken = await AuthenticationToken.getAuthenticationTokenIOS()
                if (!authToken?.authenticationToken) {
                    throw new Error("No authentication token returned from Facebook")
                }
                idToken = authToken.authenticationToken
            } else {
                // On Android, use AccessToken
                // This token can be used to access the Graph API
                const accessToken = await AccessToken.getCurrentAccessToken()
                if (!accessToken?.accessToken) {
                    throw new Error("No access token returned from Facebook")
                }
                idToken = accessToken.accessToken
            }

            if (mode === "loginModeSignInToExistingAccount" || !store.user) {
                console.log(`Signing in with Facebook`, { idToken, nonce })
                // Sign in case
                const { data, error } =
                    Platform.OS === "ios"
                        ? await db.auth.signInWithIdToken({
                              provider: "facebook",
                              token: idToken,
                               //on iOS pass the raw nonce, gotrue will hash it on server side and compare to nonce in idToken
                              nonce: nonce!
                          })
                        //Warning: this does not work on android because the android facebook sdk won't give you a standard jwt token
                       // it will only give you an access_token with which you can call facebook graph apis
                        : await db.auth.signInWithIdToken({
                              provider: "facebook",
                              token: idToken
                          })

                if (error) throw error
                log.info("Successfully signed in with Facebook", data)
            } 
```

**Reply by eybel** [reply]
@evelant So the idea is to has it before using it for OS, I think I tried that but wasnt sure if I did it right. I will work on this soon again, so far I. left it SKIP NONCE to true. Thank you!

**Reply by chanphiromsok** [reply]
> @hugoh59 @ayazalphasquad @adamsucharda99 @eybel I figured out the problem. The issue is that gotrue (supabase auth) decodes the id token then compares the nonce from the token to _the hash of the nonce you pass in to signInWithIdToken_. That means you need to hash your nonce, pass the hashed nonce to the native sign in method, then pass the raw nonce to `signInWithIdToken`. That way gotrue will hash the raw nonce and compare it to the nonce in the token and they will match. Example:
> 
> ```ts
>   // Generate nonce for iOS limited login
>             const nonce = Platform.OS === "ios" ? uuid() : undefined
>             //need to hash the nonce for iOS limited login since supabase checks the nonce against the hashed nonce not the original
>             const hashedNonce = await digestStringAsync(
>                 CryptoDigestAlgorithm.SHA256,
>                 nonce ?? "" // Provide empty string as fallback
>             )
>             // Request permissions with limited tracking on iOS
>             const result = await LoginManager.logInWithPermissions(
>                 ["email", "openid"],
>                 Platform.OS === "ios" ? "limited" : undefined,
>                 // give the hashed nonce so it ends up in the token and supabase can match it by hashing the raw nonce in gotrue
>                 Platform.OS === "ios" ? hashedNonce : undefined
>             )
>             console.log(`Facebook login result`, result)
>             if (result.isCancelled) {
>                 throw new Error("Facebook login cancelled")
>             }
> 
>             // Get the authentication token
>             let idToken: string
>             if (Platform.OS === "ios") {
>                 // On iOS, use AuthenticationToken for limited login
>                 // This token cannot be used to access the Graph API
>                 const authToken = await AuthenticationToken.getAuthenticationTokenIOS()
>                 if (!authToken?.authenticationToken) {
>                     throw new Error("No authentication token returned from Facebook")
>                 }
>                 idToken = authToken.authenticationToken
>             } else {
>                 // On Android, use AccessToken
>                 // This token can be used to access the Graph API
>                 const accessToken = await AccessToken.getCurrentAccessToken()
>                 if (!accessToken?.accessToken) {
>                     throw new Error("No access token returned from Facebook")
>                 }
>                 idToken = accessToken.accessToken
>             }
> 
>             if (mode === "loginModeSignInToExistingAccount" || !store.user) {
>                 console.log(`Signing in with Facebook`, { idToken, nonce })
>                 // Sign in case
>                 const { data, error } =
>                     Platform.OS === "ios"
>                         ? await db.auth.signInWithIdToken({
>                               provider: "facebook",
>                               token: idToken,
>                                //on iOS pass the raw nonce, gotrue will hash it on server side and compare to nonce in idToken
>                               nonce: nonce!
>                           })
>                         //Warning: this does not work on android because the android facebook sdk won't give you a standard jwt token
>                        // it will only give you an access_token with which you can call facebook graph apis
>                         : await db.auth.signInWithIdToken({
>                               provider: "facebook",
>                               token: idToken
>                           })
> 
>                 if (error) throw error
>                 log.info("Successfully signed in with Facebook", data)
>             } 
> ```


thank you this one work 
- expo ,supabase , react-native-fbsdk-next
```javascript
import { CryptoDigestAlgorithm, digestStringAsync, randomUUID } from "expo-crypto";

const getFacebookAccessToken = async () => {
  if (isAndroid) {
    const token = await AccessToken.getCurrentAccessToken();
    console.log("Facebook Access Token: Android", token);
    return token?.accessToken;
  }
  const token = await AuthenticationToken.getAuthenticationTokenIOS();
  console.log("Facebook Access Token: IOS", token);
  return token?.authenticationToken;
};
const getNonce = async () => {
    // nonce goes to Supabase's signInWithIdToken()
    const nonce = randomUUID();
    const nonceDigest = await digestStringAsync(CryptoDigestAlgorithm.SHA256, nonce);
    return { nonce, nonceDigest };
  };
  const registerFacebookWithSupabase = async () => {
    try {
      //https://github.com/orgs/supabase/discussions/22297
      const nonce = await getNonce();
      console.log("Generated Nonce:", nonce);
      const result = await LoginManager.logInWithPermissions(
        ["public_profile", "email"],
        "limited",
        isIOS ? nonce.nonceDigest : undefined // Use the nonce digest for security
      );
      console.log("Facebook Login Result:", result);
      if (result.isCancelled) {
        throw new Error("Cancelled the login process");
      }
      const accessToken = await getFacebookAccessToken();
      if (!accessToken) {
        throw new Error("Failed to get access token");
      }
      const { data, error } = await Supabase.auth.signInWithIdToken(
        isIOS
          ? {
              provider: "facebook",
              token: accessToken,
              nonce: nonce.nonce
            }
          : {
              provider: "facebook",
              token: accessToken
            }
      );

      if (!data.session) {
        throw error;
      }

      const { data: resp } = await UserService.loginWithSocial({
        email: data.user.user_metadata.email,
        password: "",
        jwt: data.session?.access_token
      });

      return resp;
    } 
  };
  ```
  
  currently work only IOS android still error
